### PR TITLE
Fix file permissions template when checking directories and allowing stricter permissions

### PR DESCRIPTION
**Pull request by ggbecker**
#### Description:

- Update1: Fixed the template to properly assess directory permissions

### CODE REVIEW

**Comment by JAORMX** [comment]
Well, CI seems happy

**Comment by JAORMX** [comment]
/test e2e-aws-ocp4-moderate

The test passed but the pod that uploads logs (provided by the CI infra) failed.

**Comment by yuumasato** [comment]
I think the issue is actually on the scanner side, see linked PR in OpenSCAP.

**Comment by JAORMX** [comment]
@yuumasato should we then merge https://github.com/ComplianceAsCode/content/pull/6706 and only merge that default when the scanner fix lands?

**Comment by yuumasato** [comment]
> @yuumasato should we then merge #6706 and only merge that default when the scanner fix lands?

Either that, or we merge #6709, as it seems to provide a reasonable workaround.

**Comment by JAORMX** [comment]
> > @yuumasato should we then merge #6706 and only merge that default when the scanner fix lands?
> 
> Either that, or we merge #6709, as it seems to provide a reasonable workaround.

@yuumasato @ggbecker whatever you guys think is best.

File: shared/templates/file_permissions/oval.template
Comment (yuumasato) [inline shared/templates/file_permissions/oval.template:39]: The use of `filter action=include` here is not wrong, but may require some explanation as to why we go the approach of "defining that the object with a certain state doesn't exist" opposed "defining the object and checking that it doesn't match a state".

**Review by yuumasato — CHANGES_REQUESTED** [review]

**Comment by yuumasato** [comment]
> @yuumasato @ggbecker whatever you guys think is best.

I'd go with this PR.
The use of `filter action="include"` enables OpenSCAP < 1.3.5 to use the file_permissions template with `allow stricter permissions`.

Revertint the template to `allow_stricter_permissions: false` won't make it work for rules who enable it.

File: shared/templates/file_permissions/oval.template
Comment (yuumasato) [reply shared/templates/file_permissions/oval.template:39]: The filter should be kept until 1.3.5 is widely deployed, but there is no harm in keeping the filter indefinitely.

File: shared/templates/file_permissions/oval.template
Comment (ggbecker) [reply shared/templates/file_permissions/oval.template:24]: I would rather remove this condition now that we know this should  work for openscap >= 1.3.5

File: shared/templates/file_permissions/oval.template
Comment (ggbecker) [reply shared/templates/file_permissions/oval.template:39]: I'll add a comment. I'll also try to unify this filter with the existing one that is being used for the file option (when the check is configured to look for a file)

**Comment by ggbecker** [comment]
> > @yuumasato @ggbecker whatever you guys think is best.
> 
> I'd go with this PR.
> The use of `filter action="include"` enables OpenSCAP < 1.3.5 to use the file_permissions template with `allow stricter permissions`, and will work after the fix is in the wild too.
> 
> Revertint the template to `allow_stricter_permissions: false` won't make it work for rules who enable it.

I've updated the pull requests with some improvements to the code structure and applied suggestions. I've also switched to read to review, please take a look again. Commits that enabled/disabled the `allow_stricter_permissions` were stripped away from the PR.

**Comment by ggbecker** [comment]
@yuumasato please advise if e87eb6b is the best solution for:

https://github.com/ComplianceAsCode/content/runs/2130339562?check_suite_focus=true#step:5:141

`oval:ssg-test_file_permissions_backup_etc_group:tst:1 - No state should be referenced when check_existence has a value of 'none_exist'.`

File: shared/templates/file_permissions/oval.template
Comment (yuumasato) [inline shared/templates/file_permissions/oval.template:7]: I think this should be `any_exist`.
If there are no files/directories with extra permissions bits set, we are fine.

File: shared/templates/file_permissions/oval.template
Comment (yuumasato) [inline shared/templates/file_permissions/oval.template:4]: Text suggestion:
`looks for files/directories that have any extra permission bit set.`

**Review by yuumasato — COMMENTED** [review]
> @yuumasato please advise if [e87eb6b](https://github.com/ComplianceAsCode/content/commit/e87eb6b3ba5ca0e4d3aadcc1f56975b442db01be) is the best solution for:

This looks to be most future proof and easier to update later, :)

With OpenSCAP < 1.3.5, the filter will include the objects we want (files/directories with extra permission bits set), and although the State comparison is buggy, we know all objects evaluated there will have at least one Entity trespassing the mask, thus matching and working as expected.

With OpenSCAP >= 1.3.5, the Entity matching will work correctly and we can just remove the filter. And forgetting to remove the filter will cause no harm.

File: shared/templates/file_permissions/oval.template
Comment (yuumasato) [reply shared/templates/file_permissions/oval.template:4]: Extra suggestion: `The permissions (STATEMODE) is negated (inverted?) and becomes a mask for permissions that should not be set.`

File: shared/templates/file_permissions/oval.template
Comment (ggbecker) [reply shared/templates/file_permissions/oval.template:7]: If we use `any_exist` every check would fail. For example, if no object is collected (there are no files/directories with wrong permissions), this check would pass (`any_exist` allows) and subsequentially fail because the criterion has `negate=true`. That means, we should keep the `at_least_one_exists` to cover this case.

Comment (yuumasato) [reply shared/templates/file_permissions/oval.template:7]: You are right, the `negate` confused me, :)

**Review by yuumasato — APPROVED** [review]
Nice fix @ggbecker :)

LGTM, just pending results from OCP rules.

**Review by JAORMX — APPROVED** [review]
Thanks for taking a look at this folks! Let's wait for CI now.

**Comment by JAORMX** [comment]
@ggbecker yeah... the failure is unrelated. It was the cluster logging stack failing to install (yet another operator).

**Comment by JAORMX** [comment]
ci/prow/e2e-aws-ocp4-cis-node passed, which was the affected job. Merging.

**Pull request merged**
